src/main/java/com/teragrep/pth_06/planner/offset/KafkaOffset.java

```diff
@@ -67,18 +67,29 @@ public class KafkaOffset extends Offset implements Serializable {
     }.getType();
 
     private final Map<String, Long> serializedKafkaOffset;
+    private final boolean stub;
+
+    public KafkaOffset() {
+        this(new HashMap<>(), true);
+    }
 
     public KafkaOffset(Map<TopicPartition, Long> offset) {
+        this(offset, false);
+    }
+
+    public KafkaOffset(Map<TopicPartition, Long> offset, boolean stub) {
         serializedKafkaOffset = new HashMap<>(offset.size());
         for (Map.Entry<TopicPartition, Long> entry : offset.entrySet()) {
 
             serializedKafkaOffset.put(entry.getKey().toString(), entry.getValue()); // offset
         }
+        this.stub = stub;
     }
 
     public KafkaOffset(String s) {
         Gson gson = new Gson();
```

Comment:
could this be moved up and use primary constructor with `this`?

Reply:
Map<TopicPartition, Long> is not serializable which is why it is converted to serializable Map<String, Long> in the constructor. And because of this logic being inside the constructors the KafkaOffset(String s) which is used by serialization testing can't be simply refactored to a secondary constructor.
The code must be refactored so the conversion to serializable map is done before the object is initialized, that way the constructor can be refactored into a secondary constructor.
Solved identical issue of HdfsOffset in commit:  b928695de1a73c144c5b4b5693cc0da52423f600
Solved in commit: cc07931a628a608264dd48ddf7d6a5c5bfe5938b

Reply:
Could this potentially use a decorator object or similar?

Reply:
The class extends the abstract class Offset and implements the interface Serializable.  AFAIK, because the class is serializable it can't encapsulate the Map<TopicPartition, Long> and the same goes for decorators.

It would be possible to use transient keyword with the Map<TopicPartition, Long> which makes serialization ignore the class field. This way Map<TopicPartition, Long> could be used as input parameter for the constructor, and a new method can be added for converting the Map<TopicPartition, Long> into a Map<String, Long> before the serialization is done. That way at least the code duplication for converting the map to serializable format would be solved.